### ISSUE TITLE
etcd snapshot restore on HA rancher cluster fails to bring up calico and Invalid bearer token

### ISSUE THREAD

**Issue by gdmello** [rancher/rke#1362]
**RKE version:**
v0.2.1

**Docker version: (`docker version`,`docker info` preferred)**
```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-45-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 15.67GiB
Name: rms-m2-dev-ft.domain.com
ID: KS6U:54P2:D7IO:NLJE:2UVW:IJCA:DOJW:5N74:6YCN:IX73:DRWV:DOVF
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```
**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
$ uname -r
4.4.0-45-generic
```
**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
VM

**Steps to Reproduce:**
1-Take an etcd snapshot (or a local backup):
```
rms-m1-dev-ft:~$ sudo rke etcd snapshot-save --config=rke-cluster-config.yaml --name rancher-backup-may-2019
INFO[0000] Starting saving snapshot on etcd hosts 
...
INFO[0016] Finished saving snapshot [rancher-backup-may-2019] on all etcd hosts 
rms-m1-dev-ft:~$ ls -atl /opt/rke/etcd-snapshots/ | grep -i 2019
total 337988
drwxr-xr-x 2 root root      276 May 23 14:38 .
-rw-r--r-- 1 root root 49434656 May 23 14:38 rancher-backup-may-2019
```
2-Restore the etcd cluster from a snapshot, on the same 3 nodes which were running the etcd nodes along with the control plane and worker roles.
```
rms-m1-dev-ft:~$ sudo rke etcd snapshot-restore --config=rke-cluster-config.yaml --name rancher-backup-may-2019
INFO[0000] Restoring etcd snapshot rancher-backup-may-2019 
INFO[0000] Successfully Deployed state file at [./rke-cluster-config.rkestate] 
…
INFO[0138] Finished building Kubernetes cluster successfully 
INFO[0138] Restarting network, ingress, and metrics pods 
INFO[0146] Finished restoring snapshot [rancher-backup-may-2019] on all etcd hosts 
```
**Results:**
The etcd cluster comes up, but the apiserver logs indicate it is not healthy:
```
E0524 06:29:42.097119       1 memcache.go:134] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0524 06:30:01.234072       1 authentication.go:62] Unable to authenticate the request due to an error: [invalid bearer token, [invalid bearer token, square/go-jose: error in cryptographic primitive]]
```
while the calico pod exits on start up as it's unable to authenticate:
```
$ docker ps -a | grep -i calico
2ed23386891b        7eca10056c8e                         "start_runit"            12 seconds ago      Exited (1) 11 seconds ago                       k8s_calico-node_calico-node-k66hz_kube-system_7a21830d-70de-11e9-8851-525400c70544_242
ed92837347fa        9f355e076ea7                         "/install-cni.sh"        20 hours ago        Up 20 hours                                     k8s_install-cni_calico-node-k66hz_kube-system_7a21830d-70de-11e9-8851-525400c70544_0
d80b9a6d03fe        rancher/pause-amd64:3.1              "/pause"                 20 hours ago        Up 20 hours                                     k8s_POD_calico-node-k66hz_kube-system_7a21830d-70de-11e9-8851-525400c70544_0
rms-m1-dev-ft:~$ docker logs 2ed
2019-05-24 15:41:31.211 [INFO][9] startup.go 251: Early log level set to info
2019-05-24 15:41:31.211 [INFO][9] startup.go 267: Using NODENAME environment for node name
2019-05-24 15:41:31.211 [INFO][9] startup.go 279: Determined node name: rms-m1-dev-ft.domain.com
2019-05-24 15:41:31.213 [INFO][9] startup.go 302: Checking datastore connection
2019-05-24 15:41:31.224 [WARNING][9] startup.go 314: Connection to the datastore is unauthorized
2019-05-24 15:41:31.224 [WARNING][9] startup.go 1058: Terminating
```
**Expected:**
The new etcd cluster is restored from the snapshot, and the entire cluster is healthy i.e. able to communicate to etcd successfully.

**Comment by gdmello** [comment]
All of these 3 nodes were Virtual Machines (VMs), and the first one on rebooting wouldn't boot up and had a corrupted disk.

With a brand new VM replacement for this problem VM, I was able to successfully restore the etcd cluster from a snapshot.

So this is a non-issue.

**Comment by verikca** [comment]
I had exact same symptoms but my situation and fix was different.

I was restoring a cluster running in OpenStack with RKE 0.2.6 (and running Rancher 2.2.6). The cluster was originally created with RKE 0.2.3 (and running Rancher 2.2.3). At one point when upgrading Rancher on this cluster we did bump into the issue described in https://github.com/rancher/rancher/issues/20699

So right now, when doing a restore we did face the same issue again, only on RKE level. What was happening, the cluster.rkestate file we had was from RKE 0.2.3 and it had the following lines inside (two occurrences):
```
"monitor-delay": 0,
"monitor-timeout": 0,
```
These are illegal values for the new RKE so RKE was simply ignoring my cluster.rkestate file replacing it with a new one and generating new certificates. Unfortunately the new certificates were not able to decrypt the old etcd database causing the exact same symptoms the original poster was having.

The fix for me was to change those values in cluster.rkestate file (make sure you use the original one) to:
```
"monitor-delay": "",
"monitor-timeout": "",
```
(two occurrences)

Then RKE etcd snapshot-restore succeeded and all went well.
